src/Agent.Plugins/PipelineArtifact/FilePathServer.cs

```diff
@@ -0,0 +1,89 @@
+            // the artifact should get cleaned up during retention even if the copy fails in the
+            // middle.
+            Directory.CreateDirectory(artifactPath);
+
+            VssConnection connection = context.VssConnection;
+            BuildServer buildServer = new BuildServer(connection);
+            var propertiesDictionary = new Dictionary<string, string>
+            {
+                { FileShareArtifactUploadEventProperties.ArtifactName, artifactName },
+                { FileShareArtifactUploadEventProperties.ArtifactType, PipelineArtifactConstants.FileShareArtifact },
+                { FileShareArtifactUploadEventProperties.ArtifactLocation, fileSharePath }
+            };
+
+            // Associate the pipeline artifact with an build artifact.
+            var artifact = await buildServer.AssociateArtifactAsync(projectId, buildId, artifactName, ArtifactResourceTypes.FilePath, fileSharePath, propertiesDictionary, token);
+            var parallel = context.GetInput(FileShareArtifactUploadEventProperties.Parallel, required: false);
+            var parallelCount = parallel == "true" ? GetParallelCount(context, context.GetInput(FileShareArtifactUploadEventProperties.ParallelCount, required: false)) : 1;
+
+            if (System.IO.Directory.Exists(fileSharePath))
+            {
+                FileShareProvider provider = new FileShareProvider(context, new CallbackAppTraceSource(str => context.Output(str), System.Diagnostics.SourceLevels.Information));
+                await provider.PublishArtifactAsync(targetPath, artifactPath, parallelCount, token);
+                context.Output(StringUtil.Loc("CopyFileComplete", artifactPath));
+            }
+        }
+
+        internal static class FileShareArtifactUploadEventProperties
+        {
+            public const string ArtifactName = "artifactname";
+            public const string ArtifactLocation = "artifactlocation";
+            public const string ArtifactType = "artifacttype";
+            public const string ParallelCount = "parallelCount";
+            public const string Parallel = "parallel";
+        }
+
+        // Enter the degree of parallelism, or number of threads used, to perform the copy. The value must be at least 1 and not greater than 128.
+        // This is the same logic as the build artifact tasks https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts
+        private int GetParallelCount(AgentTaskPluginExecutionContext context, string parallelCount)
+        {
+            var result = 8;
+            if(int.TryParse(parallelCount, out result))
+            {
+                if(result < 1) 
+                {
+                    context.Output(StringUtil.Loc("UnexpectedParallelCount"));
+                    result = 1;
+                }
+                else if(result > 128)
+                {
+                    context.Output(StringUtil.Loc("UnexpectedParallelCount"));
```

Comment:
>UnexpectedParallelCount [](start = 51, length = 23)

Just that error message is not helpful. Tell the end users what range is allowed

Reply:
This is using the loc file.. which means the string you see here is just the key. You will see the full message from the string file.

Reply:
I see that in the .json file you have the description so let's ignore this.

---
In reply to: [319295274](https://github.com/microsoft/azure-pipelines-agent/pull/2406#discussion_r319295274) [](ancestors = 319295274)